compiler/rustc_hir_analysis/src/collect/type_of/opaque.rs

```diff
@@ -138,6 +138,15 @@ impl TaitConstraintLocator<'_> {
                 continue;
             }
             constrained = true;
+            if !self.tcx.opaque_types_defined_by(item_def_id).contains(&self.def_id) {
+                self.tcx.sess.emit_err(TaitForwardCompat {
+                    span: hidden_type.span,
+                    item_span: self
+                        .tcx
+                        .def_ident_span(item_def_id)
+                        .unwrap_or_else(|| self.tcx.def_span(item_def_id)),
+                });
+            }
```

Comment:
This is the main change of this PR. We double check that any item that produces hidden types also has the corresponding opaque type in its signature